chisel/src/result.rs

```diff
@@ -0,0 +1,221 @@
+//! driver performed any transformations.
+
+use std::error::Error;
+use std::fmt::{self, Display};
+use std::fs::write;
+use std::path::PathBuf;
+
+use crate::fail;
+
+use libchisel::ModuleError;
+
+use ansi_term::Colour::{Green, Red, Yellow};
+use hex::encode;
+use parity_wasm::elements::{serialize, serialize_to_file, Module};
+
+#[derive(Clone)]
+/// Main result structure returned by ChiselDriver, containing a manifest of modules executed and
+/// exposing methods to write output from translators and creators.
+pub struct ChiselResult(Vec<RSResult>);
+
+#[derive(Clone)]
+/// Ruleset result structure.
+pub struct RSResult {
+    rs_name: String,
+    results: Vec<ModuleResult>,
+    output_path: PathBuf,
+    output_module: Option<Module>,
+}
+
+// TODO: better result handling for creators
+#[derive(Clone)]
+/// Individual module execution result. Left-hand field is the module name, and left-hand is the
+/// return value.
+pub enum ModuleResult {
+    Creator(String, Result<bool, ModuleError>),
+    Translator(String, Result<bool, ModuleError>),
+    Validator(String, Result<bool, ModuleError>),
+}
+
+impl ChiselResult {
+    pub fn new() -> Self {
+        ChiselResult(Vec::new())
+    }
+
+    pub fn rulesets_mut(&mut self) -> &mut Vec<RSResult> {
+        &mut self.0
+    }
+
+    pub fn rulesets(&self) -> &Vec<RSResult> {
+        &self.0
+    }
+}
+
+impl RSResult {
+    pub fn new(name: String) -> Self {
+        RSResult {
+            rs_name: name,
+            results: Vec::new(),
+            output_path: PathBuf::new(),
+            output_module: None,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        self.rs_name.as_str()
+    }
+
+    pub fn results_mut(&mut self) -> &mut Vec<ModuleResult> {
+        &mut self.results
+    }
+
+    pub fn set_output_path(&mut self, path: PathBuf) {
+        self.output_path = path;
+    }
+
+    pub fn set_output_module(&mut self, module: &Module) {
+        self.output_module = Some(module.clone());
+    }
+
+    pub fn write_output_to_file(&self) -> Result<bool, Box<dyn Error>> {
+        match self.output_path.to_str().expect("Output path not valid unicode") {
+            "/dev/stdout"
+                | "/dev/stderr" 
+                | "-" =>
+            {
+                fail(1, "writing raw binary to stdout or stderr is prohibited; please use a text encoded output mode.")
+            },
+            _ => (),
+        }
+        if let Some(ref module) = self.output_module {
+            // TODO: Get rid of module clone here
+            match serialize_to_file(&self.output_path, module.clone()) {
+                Ok(()) => Ok(true),
+                Err(e) => Err(e.into()),
+            }
+        } else {
+            Ok(false)
+        }
+    }
+
+    pub fn write_output_to_file_hex(&self) -> Result<bool, Box<dyn Error>> {
+        if let Some(ref module) = self.output_module {
+            // TODO: Get rid of module clone here
```

Comment:
If you want to get rid of the clone change `write_output_file_hex` to not take a ref.